packages/ui/Scrollable/Scrollbar.js

```diff
@@ -29,165 +44,130 @@ const setCSSVariable = (element, variable, value) => {
-	ignoreMode = false
-
-	update = (bounds) => {
-		const
-			{vertical} = this.props,
-			{clientWidth, clientHeight, scrollWidth, scrollHeight, scrollLeft, scrollTop} = bounds,
-			clientSize = vertical ? clientHeight : clientWidth,
-			scrollSize = vertical ? scrollHeight : scrollWidth,
-			scrollOrigin = vertical ? scrollTop : scrollLeft,
-
-			thumbSizeRatioBase = (clientSize / scrollSize),
-			scrollThumbPositionRatio = (scrollOrigin / (scrollSize - clientSize)),
-			scrollThumbSizeRatio = Math.max(this.minThumbSizeRatio, Math.min(1, thumbSizeRatioBase));
-
-		setCSSVariable(this.thumbRef.current, '--scrollbar-size-ratio', scrollThumbSizeRatio);
-		setCSSVariable(this.thumbRef.current, '--scrollbar-progress-ratio', scrollThumbPositionRatio);
-	}
-
-	showThumb = () => {
-		this.hideThumbJob.stop();
-		ReactDOM.findDOMNode(this.thumbRef.current).classList.add(this.props.css.thumbShown); // eslint-disable-line react/no-find-dom-node
-	}
-
-	startHidingThumb = () => {
-		this.hideThumbJob.start();
-	}
-
-	hideThumb = () => {
-		ReactDOM.findDOMNode(this.thumbRef.current).classList.remove(this.props.css.thumbShown); // eslint-disable-line react/no-find-dom-node
-	}
-
-	hideThumbJob = new Job(this.hideThumb.bind(this), thumbHidingDelay);
-
-	calculateMetrics = () => {
-		const primaryDimenstion = this.props.vertical ? 'clientHeight' : 'clientWidth';
-		let trackSize;
+const ScrollbarBase = forwardRef((props, ref) => {
+	// Refs
+	const containerRef = useRef();
+	const thumbRef = useRef();
+	// Job
+	const hideThumbJob = useHidingThumbJob(hideThumb, thumbHidingDelay);
+	// Render
+	const
+		{childRenderer, className, corner, css, vertical, ...rest} = props,
+		containerClassName = classNames(
+			className,
+			css.scrollbar,
+			corner && css.corner,
+			vertical ? css.vertical : css.horizontal
+		);
 
-		if (this.props.clientSize) {
-			trackSize = this.props.clientSize[primaryDimenstion];
-		} else {
-			trackSize = this.containerRef.current[primaryDimenstion];
+	delete rest.clientSize;
+
+	function hideThumb () {
+		addRemoveClass(thumbRef.current, css.thumbShown, 'remove');
+	}
+
+	useImperativeHandle(ref, () => ({
+		getContainerRef: () => (containerRef),
+		showThumb: () => {
+			hideThumbJob.stop();
+			addRemoveClass(thumbRef.current, css.thumbShown, 'add');
+		},
+		startHidingThumb: () => {
+			hideThumbJob.start();
+		},
+		update: (bounds) => {
+			const
+				{clientSize} = props,
+				primaryDimenstion = vertical ? 'clientHeight' : 'clientWidth',
+				trackSize = clientSize ? clientSize[primaryDimenstion] : containerRef.current[primaryDimenstion],
+				scrollViewSize = vertical ? bounds.clientHeight : bounds.clientWidth,
+				scrollContentSize = vertical ? bounds.scrollHeight : bounds.scrollWidth,
+				scrollOrigin = vertical ? bounds.scrollTop : bounds.scrollLeft,
+				thumbSizeRatioBase = (scrollViewSize / scrollContentSize),
+				scrollThumbPositionRatio = (scrollOrigin / (scrollContentSize - scrollViewSize)),
```

Comment:
`minThumbSizeRatio = ri.scale(minThumbSize) / trackSize;` code might be needed here before using it.

Reply:
You might miss it.

Reply:
Oh I misunderstood it. Please never mind my comments above.

By the way, could we use `React.useMemo` for the complex calculation above?
Ref: https://reactjs.org/docs/hooks-reference.html#usememo

Reply:
Good point. If this API returns _already computed_ values frequently, `useMemo` can help this to be optimized.
And, IMO, this API is not so good to use `useMemo` with following reasons;
- This API is called on every scrolling by a browser to update thumb, and the scroll position can be *any number* between 0 and the maximum scroll position. Since it looks that there is no limitation to memoize values for now, memoized values can be a set of thousands or more.
- Even if there is a limitation, we cannot expect optimization in general except some frequently positioned cases like the beginning, the ending, and some scroll positions by 5-way spotlight moving.
- This API is called only the scroll position is changed in Scrollables. When just showing and hiding a thumb, this API is not called at all.